text/0000-default-field-values.md

```diff
@@ -0,0 +1,1979 @@
+  have no limits to their complexity other than being computable,
+  these expressions are evaluated at *compile time*.
+  Thus, *`const` expressions cannot have unbounded complexity at run-time*.
+  At most, `const` expressions can create huge arrays and similar cases;
+
+  Ensuring that `Foo { .. }` remains relatively cheap is therefore important
+  because there is a general expectation that literal expressions have a small
+  and predictable run-time cost and are trivially predictable.
+  This is particularly important for Rust since this is a language that aims
+  to give a high degree of control over space and time as well as predictable
+  performance characteristics.
+
++ Keeping default values limited to `const` expressions ensures that if
+  the following situation develops:
+
+  ```rust
+  // Crate A:
+  pub struct Foo {
+      bar: u8 = const_expr,
+  }
+
+  // Crate B:
+  const fn baz() -> Foo {
+      Foo { .. }
+  }
+  ```
+
+  then crate A cannot suddenly, and unawares, cause a semver breakage
+  for crate B by replacing `const_expr` with `non_const_expr` since
+  the compiler would reject such a change (see lemmas 1-2).
+  Thus, enforcing constness gives a helping hand in respecting semantic version.
+
+  Note that if Rust would ever gain a mechanism to state that a
+  function will not diverge, e.g.:
+
+  ```rust
+  nopanic fn foo() -> u8 { 42 } // The weaker variant; more easily attainable.
+  total fn bar() -> u8 { 24 } // No divergence, period.
+  ```
+
+  then the same semver problem would manifest itself for those types of
+  functions. However, Rust does not have any such enforcement mechanism
+  right now and if it did, it is generally harder to ensure that a function
+  is total than it is to ensure that it is deterministic; thus, while
+  it is regrettable, this is an acceptable trade-off.
+
++ Finally, note that `const fn`s, when coupled with planned future extensions,
+  can become quite expressive. For example, it is already planned that `loop`s,
+  `match`es, `let` statements, and `panic!(..)`s will be allowed.
+  Another feasible extension in the future is allocation.
+
+  Therefore, constant expressions should be enough to satisfy most expressive
+  needs.
+
+## Instead of `Foo { ..Default::default() }`
+
+As an alternative to the proposed design is either explicitly writing out
+`..Default::default()` or extending the language such that `Foo { .. }` becomes
+sugar for `Foo { ..Default::default() }`. While the latter idea does not satisfy
+any of the [motivation] set out, the former does to a small extent.
+
+In particular, `Foo { .. }` as sugar slightly improves ergonomics.
+However, it has some notable problems:
+
++ Because it desugars to `Foo { ..Default::default() }`, it cannot be required
+  that the expression is a constant one. This carries all the problems noted in
+  the previous section on why default field values should be a `const` context.
```

Comment:
It could desugar to `Foo { .. const { Default::default() } }` instead.

... but that may run into the lack of const trait impl support. (Not sure if it does since the type is not generic here.)